### ISSUE TITLE
Support for other single-file-based local storage formats

### ISSUE THREAD

**Issue by Pieter1024** [pimutils/vdirsyncer#531]
I would like to implement a different local storage format, namely the one that [tudu](https://code.meskio.net/tudu/) uses. I like the interface of tudu more than the interface of todoman, and I would like to use vdirsyncer to sync a remote server with its storage.

I can basically do one of two things:
* Add a storage type to vdirsyncer for the XML-based format tudu uses.
* Add support for vdir-type storage to tudu.

Of these, I prefer the first because it is probably much simpler to implement, as vdirsyncer already has support for several local storage types.

If I write a pull request that does this:
* Would you be open to accepting it?
* Do you have some pointers on where to start?

**Comment by untitaker** [comment]
I'd rather recommend using task management app that is "native" to the VTODO format. Converting from VTODO to the XML-based format (and back) will probably not be lossless.

Of course I would accept a PR that implements this. You will need the `icalendar` library for parsing VTODOs (vdirsyncer's internal parser is very primitive). Please add that as new line in `extras_require` in `setup.py`. `xml.etree` from the stdlib will probably suffice for generating the tudu file.

The abstract interface is documented in `vdirsyncer/storage/base.py`. Each file in that folder (other than `base.py`) represents an actual storage type you can use. A very simple one is `memory.py` which stores all items in memory. The second most simple one is `filesystem.py`. Those are useful for getting a hang of the API you need to implement, but you will want to take a look at `singlefile.py` too.

So what you need to do is to create a subclass of `vdirsyncer.storage.base.Storage` and override all of its methods. You then also need to create a new `test_*.py` file in `tests/storage` that tests this new storage.

Finally you need to register that new storage in a table at https://github.com/pimutils/vdirsyncer/blob/ad8fc1dc75d931708c0c80596c0a72199906f6df/vdirsyncer/cli/utils.py#L39-L51

For an example PR take a look at #500, though that is unfinished.

Feel free to ask questions. We have an IRC channel on freenode called `#pimutils`. It isn't logged, even though there are no stupid questions :)

**Comment by Pieter1024** [comment]
Thank you for your swift response, and for the pointers. I'll look into this.

**Comment by Pieter1024** [comment]
I have since looked into this, and I have decided to not do this. In short: you were right, the XML-based format is not suitable for a straightforward implementation of synchronisation. More specifically, because the XML-based format does not contain the last modification timestamp for each item, it is not possible to do two-way synchronisation in a sensible way. Of course, it is possible to modify the format that tudu uses, but if we're going to modify tudu, we might as well modify it to use the native VTODO-format.

I think I will attempt to build an interactive, curses-based interface for todoman.
